onnx/shape_inference/implementation.cc

```diff
@@ -185,6 +202,35 @@ static void InferShapesImpl(
   std::unordered_map<std::string, const TensorProto*> inputDataByName;
   for (const auto& tp : g->initializer()) {
     inputDataByName[tp.name()] = &tp;
+    if (use_other_domain) {
+      continue;
+    }
+    // Consider the tensors from the initializer if all using onnx domain
+    TypeProto *initializerType = new TypeProto();
+    TypeProto_Tensor* initializerTensorType = initializerType->mutable_tensor_type();
+    // set the shape according to the initializer shape info
+    // if the shape info of initializer is not empty 
+    if (tp.dims_size() != 0) {
+      initializerTensorType->set_elem_type(tp.data_type()); // TensorProto_DataType_FLOAT
+      TensorShapeProto* shape = initializerTensorType->mutable_shape();
+      for (int i = 0 ; i < tp.dims_size(); ++i) {
+        shape->add_dim()->set_dim_value(tp.dims(i)); 
+      }
+    }
+    auto iter = valueTypesByName.find(tp.name());
+    // If it already exists in input, check input and initializer is sync
+    // use shape info from input (input has priority over initializer)
+    if (iter != valueTypesByName.end()) {
+        checkShapesAndTypes(*initializerTensorType, *valueTypesByName[tp.name()]->mutable_tensor_type());
+    }
+    // Support IR>=4: some tensors can only exist in initializer and not in input
+    // So shape_inference should make use of initializer shapes
+    // Store initializer shape info in value_info as well    
+    else if (ir_version >= 4 && opset_version >= 9){
```

Comment:
I don't understand why we need to look at the opset_version here. Technically, it is okay to use IR version > 4 with opset_version < 9. In this scenario, if we have an initializer that is not an input, we can treat it as a constant and it seems okay to execute the following.

Reply:
The previous modification without considering the opset in this PR will break some tests in pytorch (graphs converted from `torch.onnx.export`). Those graphs are using old opsets (< 9), but it seems that the ir_version is always the latest one in `torch.onnx.export`. In that case, do we still need to update the input from the initializer? I thought old opset corresponds to old ir_version.

The failure is as follows:
https://circleci.com/gh/onnx/onnx/5368?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
They are using opsets from pytorch.caffe2 domain.

Reply:
But what causes the checks to fail? Do those models have initializers that are not in the input-list? (Otherwise, the behavior should not be affected by this PR ... does that sound right?) If those models have initializers that are not in the input-list, how should we handle them? They should be handled as constants, right? If something breaks when we treat them as constants, then there must be some *other* issue ... either in the pytorch model exporter or within the onnx checker. Does that make sense?

Reply:
Looking at "test_while_cond" for example: it appears that the exporter is exporting ops (relational comparison ops) with int types, though ONNX does not have them. So, it is an invalid model, and it is correct for the checker to produce an error message. I think that we need a single general solution to the problem of incorrect ONNX models exported from these backends (for example, using some explicit list of known incorrect models from the checker in CI, or something similar).

Reply:
Agreed. In fact, these models keep failing in different conditions (https://github.com/onnx/onnx/pull/2783) because they have some opsets which do not come from ONNX. Then the solution will be same as what we discussed before as you mentioned above.

Therefore, I removed the opset version stuff and only consider the opset domain: if the model is using some opsets other than ONNX domain, the shape inference won't update the input from initializer. Besides, I will propose a PR about deprecating the CircleCI and consider to remove those tests (invalid models to ONNX). 

Thank you for the explanation @gramalingam

Reply:
I think even the opset check is not needed. The initializer semantics has nothing to do with opsets. It is related *only* to IR version. The opset check can lead to unintuitive behavior; if I simply add a new domain "my.special.domain" for one particular node-op, it will suddenly impact the inference/checking for other initializers/nodes/ops which have no connection to "my.special.domain". Thanks.

Reply:
Understood. Just removed other domain stuff.